docs/developer-docs/smart-contracts/best-practices/idempotency.mdx

```diff
@@ -0,0 +1,90 @@
+---
+keywords: [beginner, concept, best practices, developer]
+---
+
+import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";
+
+# Retries and Idempotency
+
+In the case of network or other issues, ICP clients (such as agents) that issue
+ingress update calls may become unable to determine whether their ingress
+request has been processed. For example, this can happen if the client loses its
+connection until the request status has been removed from the state tree (recall
+that ICP will remove the request from the system state tree some time after
+ingress expiry).
+
+Thus, it is important to design and/or use canister APIs such that it is possible to retry requests safely, even when the ICP doesn't provide any information as to previous request attempts. This page describes general approaches that both the canister authors and the clients can adopt to enable safe retries.
+
+## Idempotent Canister APIs
+
+We say that a canister endpoint is idempotent if it can be called multiple times without changing the result or canister state beyond the first call. Whenever an endpoint is idempotent or can be made idempotent by the developer, this provides an easy way to implement safe retries.
+
+Given an idempotent endpoint, you can implement retries by simply retrying the call until you observe a certified response (either a replied or rejected status), see the illustration below. If such a response is ever observed, one can be sure that the transaction has been executed exactly once thanks to idempotency. However, the application may not be willing to wait for a response indefinitely and a timeout could be implemented. Upon timeout, an error should be displayed to the user instructing them to wait until the latest message that has been sent has expired (ingress_expiry) and then manually check the status of the transaction. Ideally, timeouts should be rare and not occur during normal operation.
```

Comment:
```suggestion
Given an idempotent endpoint, you can implement retries by retrying the call until you observe a certified response (either a replied or rejected status), see the illustration below. If such a response is ever observed, one can be sure that the transaction has been executed exactly once thanks to idempotency. However, the application may not be willing to wait for a response indefinitely and a timeout could be implemented. Upon timeout, an error should be displayed to the user instructing them to wait until the latest message that has been sent has expired (ingress_expiry) and then manually check the status of the transaction. Ideally, timeouts should be rare and not occur during normal operation.
```

Reply:
This advice for when a timeout happens is probably the best we can do but it's also not great. The user submitted a transactions, like a token transfer. The front-end now needs to keep that in the UI as a potentially failed transfer. The user can't make another transfer since that might be a double transfer.

I'm wondering if the front end should ever timeout and stop trying to reconcile the state.

If the user goes to another device, they might not even see the attempted transaction which is in an unknown state.